CONTRIBUTING.md

```diff
@@ -0,0 +1,85 @@
+**Thanks for taking the time to contribute! ❤️**
+
+All types of contributions are encouraged and valued, no matter if it's a bug
+report 🐛, a feature request 💡, or a Pull Request 🚀.
+
+- **❓ I have a question:** [Open a Q&A Discussion]
+- **🐛 I found a bug:** [Open an Issue]
+- **💡 I have an idea:** [Open an Idea Discussion]
+- **💻 I want to code:** See below
+
+## Contributing code
+
+1. 🔀 Fork the repo
+2. 💻 Open the repo in your editor
+3. 👨‍💻 Add your changes to your workspace
+4. ✨ Run the tests to make sure everything works
+5. 🔖 Commit & push your changes
+6. 🔁 Open a PR to get your changes merged
+7. 🚀 Profit!
+
+👩‍⚖️ When contributing code to this project, you must agree that you have authored
+100% of the content, that you have the necessary rights to the content and that
+the content you contribute may be provided under the project license.
+
+## Conventions & linting
+
+This repository is sort-of like a monorepo. We use a monorepo commit convention
+where the PR title becomes the commit message of a single squashed commit when
+merged with the `main` branch.
+
+This means that your PR should at least _try_ to conform to the conventional
+commits format. Usually that means something like "feat(deno): make it work on
+M1 Macs". But don't worry, it's OK if it doesn't. 🤷‍♂️ It just helps make the
+commits easier to reason about when there's _so many subfolders_.
+
+📚 Further reading: [Conventional Commits]
+
+We also have set up [MegaLinter] for this project, so you should get a nice
+linting report as a comment on your PR. Don't let it spook you! 🦇
+
+_Originally discussed in [devcontainers-community#4]_
+
+## What qualifies as a feature?
+
+Features are generally medium-to-large blocks of software that are tooling for a
+particular ecosystem. Things like programming languages ([Deno], [C++],
+[Haskell], [Julia]) with their popular associated tooling ([pnpm], [Poetry],
+[CMake], [Gradle]), and popular standalone tools ([MySQL], [Emscripten], [act],
+[Terraform]) are good examples of things that are features. Things that aren't
+features are things like package manager items ([TypeScript], [TensorFlow],
+[Prettier]).
+
+A good rule of thumb 👍 is that if it's **installed in a folder** or via a
+**lang-chain dependency manager** (i.e. able to be installed per-project), it's
+probably **not a feature**. If it's **global**, it might be **worth a feature**.
+If you're in doubt, [pose your idea in a new Discussion]! 💡
+
+If you'd like to check whether or not your feature already exists, check out the
+[official @devcontainers community feature index].
```

Comment:
@eitsupi specifically looking for feedback here. this is @jcbhmr speaking in this text, and as such reflects my own opinions. what's your opinion on what should be a "feature" and not?

Reply:
I think it's good, but since it's not a consensus of the entire ecosystem, can we add an annotation like "Criteria for adding to this repository"?

Another interesting example is `shell-history`.
Which doesn't actually install anything but is quite popular (I use it all the time).
https://github.com/stuartleeks/dev-container-features/tree/main/src/shell-history

Reply:
> Another interesting example is shell-history.
Which doesn't actually install anything but is quite popular (I use it all the time).
https://github.com/stuartleeks/dev-container-features/tree/main/src/shell-history

I like this! 😀 this means that my previous definition is **incomplete**

> I think it's good, but since it's not a consensus of the entire ecosystem, can we add an annotation like "Criteria for adding to this repository"?

i like the "Criteria for adding to this repository" a lot! ❤️

here's another draft idea:

1. moderately popular tool
    1. can be an idea or script that is better packaged as a feature (much more case-by-case) _this is to address the shell-history -like features_
    2. lang-chain dependency managers should use `postCreateCommand: "npm install -g @devcontainers/cli"` or similar
    3. try to avoid using devcontainer.json as your requirements.txt -- use existing dependency toolchains for project-local dependencies
4. not offered in the official @devcontainers org (if there's an official version and you want to add something, open a pr!)
5. not a part of an already healthy organization (like julia's feature repo https://github.com/JuliaLang/devcontainer-features)

or we could stick with popular _tools_ only and avoid things that are like shell-history. at least for now 😜

thoughts & opinions?